JavadocMethod: check for duplicate throws annotations

I would like a new feature in `JavadocMethod` where it checks that annotations like `@throws` are unique and don't repeat the same exception.
This check verifies that the tag exists if the method declaration calls for it. I recently found a mistake in one of my JavaDocs where I duplicated a `@throws Exception` tag twice. I'm thinking this somehow happened during a rebase, and didn't notice it at the time.
Currently, the check does check for duplicate `@return` and `@param` annotations.

````
$ cat TestClass.java
public class TestClass {
    /**
    * @param a One.
    * @param a Two.
    * @return One.
    * @return Two.
    * @throws Exception One.
    * @throws Exception Two.
    */
    int method(String a) throws Exception {
        return 0;
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="JavadocMethod" />
    </module>
</module>

$ java -jar checkstyle-7.1.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:4:7: Unused @param tag for 'a'. [JavadocMethod]
[ERROR] TestClass.java:6:7: Duplicate @return tag. [JavadocMethod]
Audit done.
Checkstyle ends with 2 errors.
````

The 2 violations are correct, but I was also expecting a violation for line 8's `@throws Exception Two` as it is a duplicate of line 7 and is therefore unneeded.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java
@@ -710,24 +710,26 @@ public class JavadocMethodCheck extends AbstractJavadocCheck {
         final DetailAST parent = getRecordDef(compactDef);
         final List<DetailAST> params = getRecordComponents(parent);
 
-        final ListIterator<JavadocTag> tagIt = tags.listIterator();
-        while (tagIt.hasNext()) {
-            final JavadocTag tag = tagIt.next();
-
+        for (JavadocTag tag : tags) {
             if (!tag.isParamTag()) {
                 continue;
             }
 
-            tagIt.remove();
+            if (isDuplicateParamTag(tags, tag)) {
+                log(tag.getLineNo(), tag.getColumnNo(), MSG_DUPLICATE_TAG,
+                        JavadocTagInfo.PARAM.getText());
+            }
+            else {
+                final String arg1 = tag.getFirstArg();
+                final boolean found = removeMatchingParam(params, arg1);
 
-            final String arg1 = tag.getFirstArg();
-            final boolean found = removeMatchingParam(params, arg1);
-
-            if (!found) {
-                log(tag.getLineNo(), tag.getColumnNo(), MSG_UNUSED_TAG,
-                        JavadocTagInfo.PARAM.getText(), arg1);
+                if (!found) {
+                    log(tag.getLineNo(), tag.getColumnNo(), MSG_UNUSED_TAG,
+                            JavadocTagInfo.PARAM.getText(), arg1);
+                }
             }
         }
+        tags.removeIf(JavadocTag::isParamTag);
 
         if (!allowMissingParamTags && reportExpectedTags) {
             for (DetailAST param : params) {
@@ -752,30 +754,32 @@ public class JavadocMethodCheck extends AbstractJavadocCheck {
                 .getTypeParameters(parent);
 
         // Loop over the tags, checking to see they exist in the params.
-        final ListIterator<JavadocTag> tagIt = tags.listIterator();
-        while (tagIt.hasNext()) {
-            final JavadocTag tag = tagIt.next();
-
+        for (JavadocTag tag : tags) {
             if (!tag.isParamTag()) {
                 continue;
             }
 
-            tagIt.remove();
-
-            final String arg1 = tag.getFirstArg();
-            boolean found = removeMatchingParam(params, arg1);
-
-            if (arg1.endsWith(ELEMENT_END)) {
-                found = searchMatchingTypeParameter(typeParams,
-                        arg1.substring(1, arg1.length() - 1));
+            if (isDuplicateParamTag(tags, tag)) {
+                log(tag.getLineNo(), tag.getColumnNo(), MSG_DUPLICATE_TAG,
+                        JavadocTagInfo.PARAM.getText());
             }
+            else {
+                final String arg1 = tag.getFirstArg();
+                boolean found = removeMatchingParam(params, arg1);
 
-            // Handle extra JavadocTag
-            if (!found) {
-                log(tag.getLineNo(), tag.getColumnNo(), MSG_UNUSED_TAG,
-                        JavadocTagInfo.PARAM.getText(), arg1);
+                if (arg1.endsWith(ELEMENT_END)) {
+                    found = searchMatchingTypeParameter(typeParams,
+                            arg1.substring(1, arg1.length() - 1));
+                }
+
+                // Handle extra JavadocTag
+                if (!found) {
+                    log(tag.getLineNo(), tag.getColumnNo(), MSG_UNUSED_TAG,
+                            JavadocTagInfo.PARAM.getText(), arg1);
+                }
             }
         }
+        tags.removeIf(JavadocTag::isParamTag);
 
         // Now dump out all type parameters/parameters without tags :- unless
         // the user has chosen to suppress these problems
@@ -792,6 +796,26 @@ public class JavadocMethodCheck extends AbstractJavadocCheck {
                     + ELEMENT_END);
             }
         }
+    }
+
+    /**
+     * Checks if the {@code @param} tag duplicates an earlier {@code @param} tag.
+     *
+     * @param tags all tags to check
+     * @param tag the tag to check
+     * @return true if there is an earlier {@code @param} tag with the same argument
+     */
+    private static boolean isDuplicateParamTag(List<JavadocTag> tags, JavadocTag tag) {
+        final int tagIndex = tags.indexOf(tag);
+        boolean result = false;
+        for (int tagPosition = 0; tagPosition < tagIndex; tagPosition++) {
+            final JavadocTag currentTag = tags.get(tagPosition);
+            if (currentTag.isParamTag()
+                    && currentTag.getFirstArg().equals(tag.getFirstArg())) {
+                result = true;
+            }
+        }
+        return result;
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheckTest.java
@@ -632,6 +632,27 @@ public class JavadocMethodCheckTest extends AbstractModuleTestSupport {
 
     }
 
+    @Test
+    public void testJavadocMethodDuplicateThrows() throws Exception {
+        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
+        verifyWithInlineConfigParser(
+                getPath("InputJavadocMethodDuplicateThrows.java"), expected);
+    }
+
+    @Test
+    public void testJavadocMethodDuplicateParam() throws Exception {
+        final String[] expected = {
+            "21:8: " + getCheckMessage(MSG_DUPLICATE_TAG, "@param"),
+            "30:8: " + getCheckMessage(MSG_DUPLICATE_TAG, "@param"),
+            "38:8: " + getCheckMessage(MSG_DUPLICATE_TAG, "@param"),
+            "47:8: " + getCheckMessage(MSG_UNUSED_TAG, "@param", "extra"),
+            "48:8: " + getCheckMessage(MSG_DUPLICATE_TAG, "@param"),
+            "57:12: " + getCheckMessage(MSG_DUPLICATE_TAG, "@param"),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputJavadocMethodDuplicateParam.java"), expected);
+    }
+
     /**
      * Verifies that the check fails on unsupported Javadoc tokens.
      *

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodDuplicateParam.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodDuplicateParam.java
@@ -1,0 +1,62 @@
+/*
+JavadocMethod
+allowInlineReturn = (default)false
+allowedAnnotations = (default)Override
+validateThrows = (default)false
+accessModifiers = (default)public, protected, package, private
+allowMissingParamTags = (default)false
+allowMissingReturnTag = (default)false
+violateExecutionOnNonTightHtml = (default)false
+tokens = (default)METHOD_DEF, CTOR_DEF, ANNOTATION_FIELD_DEF, COMPACT_CTOR_DEF
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.javadocmethod;
+
+public class InputJavadocMethodDuplicateParam {
+
+    // violation 3 lines below 'Duplicate @param tag.'
+    /**
+     * @param value first description
+     * @param value second description
+     */
+    void duplicateMethodParam(String value) {
+    }
+
+    // violation 4 lines below 'Duplicate @param tag.'
+    /**
+     * @param value first description
+     * @param other other description
+     * @param value second description
+     */
+    void duplicateMethodParamSeparated(String value, String other) {
+    }
+
+    // violation 3 lines below 'Duplicate @param tag.'
+    /**
+     * @param <T> first description
+     * @param <T> second description
+     * @param value value description
+     */
+    <T> void duplicateTypeParam(T value) {
+    }
+
+    // violation 3 lines below 'Unused @param tag for 'extra'.'
+    // violation 3 lines below 'Duplicate @param tag.'
+    /**
+     * @param extra first extra description
+     * @param extra second extra description
+     */
+    void duplicateUnusedParam() {
+    }
+
+    record Compact(String value) {
+        // violation 3 lines below 'Duplicate @param tag.'
+        /**
+         * @param value first description
+         * @param value second description
+         */
+        Compact {
+        }
+    }
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodDuplicateThrows.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/javadocmethod/InputJavadocMethodDuplicateThrows.java
@@ -1,0 +1,59 @@
+/*
+JavadocMethod
+allowInlineReturn = (default)false
+allowedAnnotations = (default)Override
+validateThrows = (default)false
+accessModifiers = (default)public, protected, package, private
+allowMissingParamTags = (default)false
+allowMissingReturnTag = (default)false
+violateExecutionOnNonTightHtml = (default)false
+tokens = (default)METHOD_DEF, CTOR_DEF, ANNOTATION_FIELD_DEF, COMPACT_CTOR_DEF
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc.javadocmethod;
+
+public class InputJavadocMethodDuplicateThrows {
+    // snippets from openjdk code
+
+    /**
+     * Sets the contents, or fails if the value was already set.
+     *
+     * @param contents value to set
+     * @throws IllegalStateException if the contents was already set
+     * @throws IllegalStateException if another operation recursively attempts to set it
+     */
+    void setOrThrow(Object contents) {
+    }
+
+    /**
+     * Closes a resource arena.
+     *
+     * @throws IllegalStateException if the arena has already been closed
+     * @throws IllegalStateException if an associated segment is being accessed concurrently
+     * @throws WrongThreadException if called from a thread other than the owner thread
+     * @throws UnsupportedOperationException if explicit close is not supported
+     * @throws RuntimeException if a custom cleanup action fails
+     */
+    void closeArena() {
+    }
+
+    /**
+     * @throws java.io.IOException first I/O failure case
+     * @throws java.io.IOException second I/O failure case
+     * @throws java.lang.NullPointerException null state failure
+     * @throws java.io.IOException fourth I/O failure case
+     */
+    void repeatedIoFailure() throws java.io.IOException {
+    }
+
+    /**
+     * Validates a format pattern.
+     *
+     * @param pattern pattern to validate
+     * @throws IllegalArgumentException if {@code pattern} is invalid
+     * @throws IllegalArgumentException if {@code pattern} is null
+     */
+    static void validateMessageFormatPattern(String pattern) {
+    }
+}

